run_tests_browser.html

```diff
@@ -0,0 +1,207 @@
+        }
+        if (val === null || val === 10) {
+          print(tty)
+        }
+      },
+      flush(tty) {
+        print(tty)
+      },
+    }
+  }
+
+  function setupStreams(FS, TTY) {
+    let mytty = FS.makedev(FS.createDevice.major++, 0)
+    let myttyerr = FS.makedev(FS.createDevice.major++, 0)
+    TTY.register(mytty, make_tty_ops())
+    TTY.register(myttyerr, make_tty_ops())
+    FS.mkdev('/dev/mytty', mytty)
+    FS.mkdev('/dev/myttyerr', myttyerr)
+    FS.unlink('/dev/stdin')
+    FS.unlink('/dev/stdout')
+    FS.unlink('/dev/stderr')
+    FS.symlink('/dev/mytty', '/dev/stdin')
+    FS.symlink('/dev/mytty', '/dev/stdout')
+    FS.symlink('/dev/myttyerr', '/dev/stderr')
+    FS.closeStream(0)
+    FS.closeStream(1)
+    FS.closeStream(2)
+    FS.open('/dev/stdin', 0)
+    FS.open('/dev/stdout', 1)
+    FS.open('/dev/stderr', 1)
+  }
+
+  self.onmessage = async ({data}) => {
+    const {files, python_code} = data
+    await importScripts('https://cdn.jsdelivr.net/pyodide/v0.21.0a3/full/pyodide.js')
+    const pyodide = await loadPyodide()
+    const {FS} = pyodide
+    setupStreams(FS, pyodide._module.TTY)
+    FS.mkdir('/test_dir')
+    FS.chdir('/test_dir')
+    await pyodide.loadPackage(['micropip', 'pytest', 'pytz'])
+    try {
+      await pyodide.runPythonAsync(python_code, {globals: pyodide.toPy({files})})
+    } catch (err) {
+      console.error(err)
+      const raw_error = Array.from(new TextEncoder().encode(err.toString()))
+      self.postMessage(raw_error)
+    }
+    post()
+  }
+</script>
+
+<script src="https://smokeshow.helpmanual.io/2d1p666q3x5f0j26223o/ansi-to-html.browser.js"></script>
+<script>
+  async function get_json(url) {
+    const cache_key = `file-cache:${url}`
+    const cache_value = localStorage.getItem(cache_key)
+    if (cache_value) {
+      return JSON.parse(cache_value)
+    }
+    const r = await fetch(url)
+    if (r.ok) {
+      let obj = await r.json()
+      localStorage.setItem(cache_key, JSON.stringify(obj))
+      return obj
+    } else {
+      const text = await r.text()
+      console.error('unexpected response', r, text)
+      throw new Error(`${r.status} ${r.statusText}`)
+    }
+  }
+
+  async function get_files(dir, tree_info) {
+    let path = `${dir}/${tree_info.path}`
+    const details = await get_json(tree_info.url)
+    if (tree_info.type === 'tree') {
+      const files = await Promise.all(details.tree.map(t => get_files(path, t)))
+      return files.flat()
+    } else {
+      return [{path, content: atob(details.content)}]
+    }
+  }
+
+  async function download_test_files() {
+    // const commits = await get_json('https://api.github.com/repos/samuelcolvin/pydantic-core/commits')
+    // const commit = await get_json(commits[0].commit.url)
+    // this commit matches the version of pydantic-core compiled below
+    const commit = await get_json('https://api.github.com/repos/samuelcolvin/pydantic-core/git/commits/515ff1004c866b23dbbd05e543c4a936ed2fdb09')
```

Comment:
![image](https://user-images.githubusercontent.com/1556054/179448282-47424b47-af6d-4059-908b-f770be393b44.png)

Unauthenticated GitHub API requests can easily get rate limited.

Reply:
Ye, but I don't want to put a github api key in the file.

It should be able to download the files once before the rate limit kicks in, files are cached after the first download. If we end up using this more, I'll zip the files and upload them.